File: yetutil/src/main/java/net/yet/util/Tick.java
package net.yet.util;

public class Tick {
	private static final String TAG = "tick";

	private long start_time = 0;
	private long end_time = 0;

	public Tick() {
		start();
	}

	/**
	 * 开始计时
	 * 
	 * @Description:
	 * @see:
	 * @since:
	 * @author: yangentao
	 * @date:2012-7-6
	 */

	public void start() {
		start_time = System.currentTimeMillis();
	}

	/**
	 * 停止计时
	 * 
	 * @Description:
	 * @see:
	 * @since:
	 * @author: yangentao
	 * @date:2012-7-5
	 */

	private void end() {
		end_time = System.currentTimeMillis();
	}

	/**
	 * 结束计时
	 * 
	 * @Description:
	 * @param prefix
	 *            打印信息
	 * @return
	 * @see:
	 * @since:
	 * @author: yangentao
	 * @date:2012-7-6
	 */
	public void end(String prefix) {
		end(0, prefix);
	}

	/**
	 * @param warnLevel
	 *            计时大于这个值(单位毫秒)时, 用警告输出
	 * @param prefix
	 */
	public void end(long warnLevel, String prefix) {
		end();
		long tick = getTick();
		if (null == prefix) {
			prefix = "";
		}
		if (warnLevel > 0 && tick > warnLevel) {
			xlog.wTag(TAG, "[TimeTick]", prefix, ":", tick, "(expect<", warnLevel, ")");
		} else {
			xlog.dTag(TAG, "[TimeTick]", prefix, ":", tick);
		}
		start();
	}

	/**
	 * 返回开始和结束的时间差
	 * 
	 * @Description:
	 * @return
	 * @see:
	 * @since:
	 * @author: yangentao
	 * @date:2012-7-6
	 */

	public long getTick() {
		return end_time - start_time;
	}

	private static Tick t = new Tick();

	/**
	 * 开始计时
	 * 
	 * @Description:
	 * @return 计时器
	 * @see:
	 * @since:
	 * @author: yangentao
	 * @date:2012-7-6
	 */
	public static Tick global() {
		return t;
	}
}

File: yetutil/src/main/java/yet/util/Tick.kt
package yet.util

import yet.util.log.xlog


/**
 * Created by yangentao on 2015/11/21.
 * entaoyang@163.com
 */
class Tick {
	val TAG = xlog.tag
	private var start_time: Long = System.currentTimeMillis()
	private var end_time: Long = 0

	fun restart(){
		start_time = System.currentTimeMillis()
	}

	fun end(prefix: String = ""): Long {
		return end(0, prefix)
	}

	fun end(warnLevel: Long, prefix: String): Long {
		end_time = System.currentTimeMillis()
		val tick = end_time - start_time
		if (warnLevel in 1..(tick - 1)) {
			xlog.wTag(TAG, "[TimeTick]", prefix, ":", tick, "(expect<", warnLevel, ")")
		} else {
			xlog.dTag(TAG, "[TimeTick]", prefix, ":", tick)
		}
		start_time = System.currentTimeMillis()
		return tick
	}
}

inline fun <R> tick(msg: String = "", block: () -> R): R {
	val t = Tick()
	val r = block()
	t.end(msg)
	return r
}

